python/ray/ml/trainer.py

```diff
@@ -210,8 +252,72 @@ def fit(self) -> Result:
             TrainingFailedError: If any failures during the execution of
             ``self.as_trainable()``.
         """
-        raise NotImplementedError
+        trainable = self.as_trainable()
+
+        from ray import tune
+        from ray.tune import TuneError
+
+        # TODO(amog/xwjiang): Replace with Tuner.
+        try:
+            analysis = tune.run(run_or_experiment=trainable, **self.run_config)
+        except TuneError:
+            raise TrainingFailedError
+        else:
+            assert len(analysis.trials) == 1
+
+            trial = analysis.trials[0]
+
+            result = Result(
+                metrics=trial.last_result,
+                checkpoint=Checkpoint.from_directory(trial.checkpoint.value)
+                if trial.checkpoint.value
+                else None,
+            )
+
+            return result
 
     def as_trainable(self) -> Type[Trainable]:
         """Convert self to a ``tune.Trainable`` class."""
-        raise NotImplementedError
+
+        base_config = self._param_dict
+        trainer_cls = self.__class__
+
+        def train_func(config, checkpoint_dir=None):
+            # config already contains merged values.
+            # Instantiate new Trainer in Trainable.
+            trainer = trainer_cls(**config)
+
+            if checkpoint_dir:
+                trainer.resume_from_checkpoint = Checkpoint.from_directory(
+                    checkpoint_dir
+                )
+
+            trainer.setup()
+            trainer.preprocess_datasets()
+            trainer.training_loop()
+
+        trainable_cls = wrap_function(train_func)
+
+        scaling_config = self.scaling_config
+
+        class TrainTrainable(trainable_cls):
+            """Add default resources to the Trainable."""
```

Comment:
This nested class is a bit unfortunate. No way to remove this or move it to Tune?

Reply:
Yes there is definitely an opportunity to clean this up! But I think let's revisit that for beta.